src/platform/ESP32/OTAImageProcessorImpl.cpp

```diff
@@ -0,0 +1,210 @@
+    {
+        ChipLogError(SoftwareUpdate, "ImageProcessor context is null");
+        return;
+    }
+    esp_err_t err = esp_ota_end(imageProcessor->mOTAUpdateHandle);
+    if (err != ESP_OK)
+    {
+        if (err == ESP_ERR_OTA_VALIDATE_FAILED)
+        {
+            ESP_LOGE(TAG, "Image validation failed, image is corrupted");
+        }
+        else
+        {
+            ESP_LOGE(TAG, "esp_ota_end failed (%s)!", esp_err_to_name(err));
+        }
+        return;
+    }
+    imageProcessor->ReleaseBlock();
+    ChipLogProgress(SoftwareUpdate, "OTA image downloaded to offset 0x%x", imageProcessor->mOTAUpdatePartition->address);
+}
+
+void OTAImageProcessorImpl::HandleAbort(intptr_t context)
+{
+    auto * imageProcessor = reinterpret_cast<OTAImageProcessorImpl *>(context);
+    if (imageProcessor == nullptr)
+    {
+        ChipLogError(SoftwareUpdate, "ImageProcessor context is null");
+        return;
+    }
+    if (esp_ota_abort(imageProcessor->mOTAUpdateHandle) != ESP_OK)
+    {
+        ESP_LOGE(TAG, "ESP OTA abort failed");
+    }
+    imageProcessor->ReleaseBlock();
+}
+
+void OTAImageProcessorImpl::HandleProcessBlock(intptr_t context)
+{
+    auto * imageProcessor = reinterpret_cast<OTAImageProcessorImpl *>(context);
+    if (imageProcessor == nullptr)
+    {
+        ChipLogError(SoftwareUpdate, "ImageProcessor context is null");
+        return;
+    }
+    else if (imageProcessor->mDownloader == nullptr)
+    {
+        ChipLogError(SoftwareUpdate, "mDownloader is null");
+        return;
+    }
+    esp_err_t err = esp_ota_write(imageProcessor->mOTAUpdateHandle, imageProcessor->mBlock.data(), imageProcessor->mBlock.size());
+    if (err != ESP_OK)
+    {
+        ESP_LOGE(TAG, "esp_ota_write failed (%s)", esp_err_to_name(err));
+        imageProcessor->mDownloader->EndDownload(CHIP_ERROR_WRITE_FAILED);
+    }
+    imageProcessor->mParams.downloadedBytes += imageProcessor->mBlock.size();
+    imageProcessor->mDownloader->FetchNextData();
+}
+
+void OTAImageProcessorImpl::HandleApply(intptr_t context)
+{
+    auto * imageProcessor = reinterpret_cast<OTAImageProcessorImpl *>(context);
+    esp_err_t err         = esp_ota_set_boot_partition(imageProcessor->mOTAUpdatePartition);
+    if (err != ESP_OK)
+    {
+        ESP_LOGE(TAG, "esp_ota_set_boot_partition failed (%s)!", esp_err_to_name(err));
+        return;
+    }
+    ESP_LOGI(TAG, "Applying, Boot partition set offset:0x%x", imageProcessor->mOTAUpdatePartition->address);
+}
+
+CHIP_ERROR OTAImageProcessorImpl::SetBlock(ByteSpan & block)
+{
+    if ((block.data() == nullptr) || block.empty())
+    {
+        return CHIP_NO_ERROR;
+    }
+
+    if (mBlock.empty())
+    {
+        mBlock = MutableByteSpan(static_cast<uint8_t *>(chip::Platform::MemoryAlloc(block.size())), block.size());
+        if (mBlock.data() == nullptr)
+        {
+            return CHIP_ERROR_NO_MEMORY;
+        }
+    }
+    CHIP_ERROR err = CopySpanToMutableSpan(block, mBlock);
```

Comment:
So if we can get called multiple times, and the incoming blocks might be different sizes, this could be a problem.  `CopySpanToMutableSpan` resizes the destination to the size of the input.  And if we get a very small block after that everything else will start failing.

Better would be to either keep track of how much storage we actually have and resize to that before doing the copy, or to free and reallocate if the destination is too small, or ideally both (in case the first block that comes through is the really small one).

Reply:
Check if the size of `block` is smaller than `mBlock`. If not, will release `mBlock` and realloc a space for its data.